ArduPlane/mode_loiter.cpp

```diff
@@ -42,28 +42,71 @@ bool ModeLoiter::isHeadingLinedUp(const Location loiterCenterLoc, const Location
     // Return true if current heading is aligned to vector to targetLoc.
     // Tolerance is initially 10 degrees and grows at 10 degrees for each loiter circle completed.
 
-    if (loiterCenterLoc.get_distance(targetLoc) < 1.05f * fabsf(plane.loiter.radius)) {
-        /* Whenever next waypoint is within the loiter radius plus 5%,
-           maintaining loiter would prevent us from ever pointing toward the next waypoint.
-           Hence break out of loiter immediately
-         */
+    // Corrected radius for altitude
+    const float loiter_radius = plane.nav_controller->loiter_radius(fabsf(plane.loiter.radius));
+    if (!is_positive(loiter_radius)) {
+        // Zero is invalid, protect against divide by zero for destination inside loiter radius case
         return true;
     }
 
-    // Bearing in centi-degrees
-    const int32_t bearing_cd = plane.current_loc.get_bearing_to(targetLoc);
-    return isHeadingLinedUp_cd(bearing_cd);
+    // Calculate relative position of the vehicle relative to loiter center projected onto the closest point of the loiter circle
+    // This removes error due to radial position as the nav controller attempts to track the circle
+    const Vector2f projected_pos = loiterCenterLoc.get_distance_NE(plane.current_loc).normalized() * loiter_radius;
+
+    // Target position relative to loiter center
+    const Vector2f target_pos = loiterCenterLoc.get_distance_NE(targetLoc);
+
+    // Distance between loiter circle and target
+    const float target_dist = target_pos.length();
+    if (!is_positive(target_dist)) {
+        // Target is coincident with loiter center, no heading will be closer than any other
+        return true;
+    }
+
+    // Target bearing in centi-degrees
+    int32_t target_bearing_cd;
+
+    if (target_dist >= loiter_radius) {
+        // Destination outside loiter radius, heading will always line up with destination
+
+        // Vector from between projected vehicle position and target postion
+        const Vector2f pos_to_target = target_pos - projected_pos;
+        target_bearing_cd = degrees(pos_to_target.angle()) * 100;
```

Comment:
I believe this should improve overshooting performance as well now that our bearing check uses a vector on the loiter circle instead of just the current vehicle location.
`const int32_t bearing_cd = plane.current_loc.get_bearing_to(targetLoc);`

Reply:
This would introduce a discontinuity as the target way-point was one or other side of the target loiter radius. We might be able to workout a similar change to the destination inside loiter case. 

I guess the other thing would be to use the radius were currently flying at rather than the target loiter radius, that would remove the need for the project position.

Reply:
Gotcha. Wouldn't this PR trade one sharp discontinuity for a less sharp, more intuitive discontinuity? Currently on master there is a sharp discontinuity with the breakout logic that exists on either side of 1.05 radius that you removed in favor of this change https://github.com/ArduPilot/ardupilot/pull/26102/files#diff-44a8229ea60a7527d6de51aa7516d83e142111a5b00c7d870f699fe01e1bcdf0L45-L49